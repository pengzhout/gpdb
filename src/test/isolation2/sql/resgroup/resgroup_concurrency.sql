-- create a resource group when gp_resource_manager is queue
0:CREATE RESOURCE GROUP rg_concurrency_test WITH (concurrency=2, cpu_rate_limit=.02, memory_limit=.02);
0:SELECT r.rsgname, num_running, num_queueing, num_queued, num_executed FROM gp_toolkit.gp_resgroup_status s, pg_resgroup r WHERE s.groupid=r.oid AND r.rsgname='rg_concurrency_test';

-- no query has been assigned to the this group
1:SELECT r.rsgname, num_running, num_queueing, num_queued, num_executed FROM gp_toolkit.gp_resgroup_status s, pg_resgroup r WHERE s.groupid=r.oid AND r.rsgname='rg_concurrency_test';
1:CREATE role role_concurrency_test RESOURCE GROUP rg_concurrency_test;
2:SET role role_concurrency_test;
2:BEGIN;
3:SET role role_concurrency_test;
3:BEGIN;
4:SET role role_concurrency_test;
4&:BEGIN;
-- new transaction will be blocked when the concurrency limit of the resource group is reached.
1:SELECT r.rsgname, num_running, num_queueing, num_queued, num_executed FROM gp_toolkit.gp_resgroup_status s, pg_resgroup r WHERE s.groupid=r.oid AND r.rsgname='rg_concurrency_test';
1:SELECT waiting_reason, rsgqueueduration > '0'::interval as time from pg_stat_activity where current_query = 'BEGIN;' and rsgname = 'rg_concurrency_test';
2:END;
3:END;
4<:
4:END;
1:SELECT r.rsgname, num_running, num_queueing, num_queued, num_executed FROM gp_toolkit.gp_resgroup_status s, pg_resgroup r WHERE s.groupid=r.oid AND r.rsgname='rg_concurrency_test';
1:DROP role role_concurrency_test;
1:DROP RESOURCE GROUP rg_concurrency_test;
