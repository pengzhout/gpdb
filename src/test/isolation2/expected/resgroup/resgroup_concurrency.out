-- create a resource group when gp_resource_manager is queue
0:CREATE RESOURCE GROUP rg_concurrency_test WITH (concurrency=2, cpu_rate_limit=.02, memory_limit=.02);
CREATE
0:SELECT r.rsgname, num_running, num_queueing, num_queued, num_executed FROM gp_toolkit.gp_resgroup_status s, pg_resgroup r WHERE s.groupid=r.oid AND r.rsgname='rg_concurrency_test';
rsgname            |num_running|num_queueing|num_queued|num_executed
-------------------+-----------+------------+----------+------------
rg_concurrency_test|0          |0           |0         |0           
(1 row)

-- no query has been assigned to the this group
1:SELECT r.rsgname, num_running, num_queueing, num_queued, num_executed FROM gp_toolkit.gp_resgroup_status s, pg_resgroup r WHERE s.groupid=r.oid AND r.rsgname='rg_concurrency_test';
rsgname            |num_running|num_queueing|num_queued|num_executed
-------------------+-----------+------------+----------+------------
rg_concurrency_test|0          |0           |0         |0           
(1 row)
1:CREATE role role_concurrency_test RESOURCE GROUP rg_concurrency_test;
CREATE
2:SET role role_concurrency_test;
SET
2:BEGIN;
BEGIN
3:SET role role_concurrency_test;
SET
3:BEGIN;
BEGIN
4:SET role role_concurrency_test;
SET
4&:BEGIN;  <waiting ...>
-- new transaction will be blocked when the concurrency limit of the resource group is reached.
1:SELECT r.rsgname, num_running, num_queueing, num_queued, num_executed FROM gp_toolkit.gp_resgroup_status s, pg_resgroup r WHERE s.groupid=r.oid AND r.rsgname='rg_concurrency_test';
rsgname            |num_running|num_queueing|num_queued|num_executed
-------------------+-----------+------------+----------+------------
rg_concurrency_test|2          |1           |1         |2           
(1 row)
1:SELECT waiting_reason, rsgqueueduration > '0'::interval as time from pg_stat_activity where current_query = 'BEGIN;' and rsgname = 'rg_concurrency_test';
waiting_reason|time
--------------+----
resgroup      |t   
(1 row)
2:END;
END
3:END;
END
4<:  <... completed>
BEGIN
4:END;
END
1:SELECT r.rsgname, num_running, num_queueing, num_queued, num_executed FROM gp_toolkit.gp_resgroup_status s, pg_resgroup r WHERE s.groupid=r.oid AND r.rsgname='rg_concurrency_test';
rsgname            |num_running|num_queueing|num_queued|num_executed
-------------------+-----------+------------+----------+------------
rg_concurrency_test|0          |0           |1         |3           
(1 row)
1:DROP role role_concurrency_test;
DROP
1:DROP RESOURCE GROUP rg_concurrency_test;
DROP
